csharp/ql/src/semmle/code/csharp/Callable.qll

```diff
@@ -206,7 +207,11 @@ class Callable extends DotNet::Callable, Parameterizable, ExprOrStmtParent, @cal
     exists(ReturnStmt ret | ret.getEnclosingCallable() = this | e = ret.getExpr())
     or
     e = this.getExpressionBody() and
-    not this.getReturnType() instanceof VoidType
+    not this.getReturnType() instanceof VoidType and
+    (
+      not this.(Modifiable).isAsync() or
+      not this.getReturnType() instanceof SystemThreadingTasksTaskClass
```

Comment:
This is not entirely correct. After C#7, it is enough to have a return type that has a `GetAwaiter` method.

Reply:
It is probably enough to require a generic type.